token/core/zkatdlog/crypto/ppm/ppm.go

```diff
@@ -59,6 +65,9 @@ func (v *PublicParamsManager) ForceFetch() error {
 }
 
 func (v *PublicParamsManager) PublicParams() *crypto.PublicParams {
+	v.mutex.Lock()
```

Comment:
What about using first a read lock to check if v.pp is not nil and then a full lock in the case the public params must be fetched?

Reply:
Is this working ? I'm not really familiar with the RWmutex 

```
	v.mutex.RLock()
	defer v.mutex.RUnlock()
	if v.pp == nil {
		v.mutex.Lock()
		defer v.mutex.Unlock()
		if v.publicParamsLoader == nil {
			panic("public parameters loaded not set")
		}
		var err error
		v.pp, err = v.publicParamsLoader.Load()
		if err != nil {
			panic(err)
		}
	}
```

Reply:
What about this?

```go	
        v.mutex.RLock()
	if v.pp != nil {
		v.mutex.RUnlock()
		return v.pp
	}
	v.mutex.RUnlock()

	v.mutex.Lock()
	defer v.mutex.Unlock()
	if v.pp == nil {
		if v.publicParamsLoader == nil {
			panic("public parameters loaded not set")
		}
		var err error
		v.pp, err = v.publicParamsLoader.Load()
		if err != nil {
			panic(err)
		}
	}
	return v.pp
```